framework/sandstone_opts.cpp

```diff
@@ -0,0 +1,1210 @@
+                    program_invocation_name, severity, name, arg, min, max);
+        }
+        print_explanation();
+    }
+
+    // non-const because this usually comes from optarg anyway
+    Integer operator()(char *arg = optarg) const
+    {
+        assert(name);
+        assert(arg);
+        assert(min <= max);
+        assert(Integer(min) == min);
+        assert(Integer(max) == max);
+
+        char *end = arg;
+        errno = 0;
+        MaxInteger parsed;
+        if constexpr (std::is_signed_v<Integer>)
+            parsed = strtoll(arg, &end, base);
+        else
+            parsed = strtoull(arg, &end, base);
+
+        if (*end != '\0' || *arg == '\0') {
+            // strtoll() did not consume the entire string or there wasn't anything to consume,
+            // so it can't be valid
+            fprintf(stderr, "%s: invalid argument for option '%s': %s\n", program_invocation_name,
+                    name, arg);
+            print_explanation();
+            exit(EX_USAGE);
+        }
+
+        // validate range
+        Integer v = Integer(parsed);
+        bool erange = (errno == ERANGE);
+        if (erange || v != parsed || v < min || v > max) {
+            print_range_error(arg);
+            if (range_mode == OutOfRangeMode::Exit)
+                exit(EX_USAGE);
+
+            if (parsed < min || (erange && parsed == std::numeric_limits<long long>::min()))
+                v = Integer(min);
+            else if (v > max || (erange && parsed == std::numeric_limits<long long>::max()))
+                v = Integer(max);
+        }
+        return v;
+    }
+};
+
+auto parse_testrun_range(const char *arg)
+{
+    int starting_test_number = 1; // One based count for user interface, not zero based
+    int ending_test_number = INT_MAX;
+
+    char *end;
+    errno = 0;
+    starting_test_number = strtoul(arg, &end, 10);
+    if (errno == 0) {
+        if (*end == '-')
+            ending_test_number = strtoul(end + 1, &end, 10);
+        else
+            errno = EINVAL;
+    }
+    if (errno != 0) {
+        fprintf(stderr, "%s: error: --test-range requires two dash separated integer args like --test-range 1-10\n",
+                program_invocation_name);
+        return EXIT_FAILURE;
+    }
+    if (starting_test_number > ending_test_number)
+        std::swap(starting_test_number, ending_test_number);
+    if (starting_test_number < 1) {
+        fprintf(stderr, "%s: error: The lower bound of the test range must be >= 1, %d specified\n",
+                program_invocation_name, starting_test_number);
+        return EXIT_FAILURE;
+    }
+    return EXIT_SUCCESS;
+}
+
+void warn_deprecated_opt(const char *opt)
+{
+    fprintf(stderr, "%s: option '%s' is ignored and will be removed in a future version.\n",
+            program_invocation_name, opt);
+}
+
+struct CmdlineParser {
+    struct ArgsPresent {
+        // endpoints
+        bool list = false;
```

Comment:
It is `false` by default, it's inconsistent with how `std::optional<>` are defined. Either initialize them both explicitly or use implicit (which is what is needed here).

Edit: see the comment above, this entire structure (`ArgsPresent`) is not needed.